cep-20.2.md

```diff
@@ -0,0 +1,651 @@
+  # unsure if we should keep this one
+  license_url: url
+
+  # REMOVED:
+  # prelink_message:
+  # license_family: string (deprecated due to SPDX)
+  # identifiers: [string]
+  # tags: [string]
+  # keywords: [string]
+  # doc_source_url: url
+```
+
+## Extra section
+
+```yaml
+# a free form dictionary
+extra:
+  <key>: <value>
+```
+
+## Example recipe
+
+What follows is an example recipe, using the YAML syntax discussed in https://github.com/conda-incubator/ceps/pull/54
+
+```yaml
+context:
+  name: xtensor
+  version: 0.24.6
+
+package:
+  name: ${{ name|lower }}
+  version: ${{ version }}
+
+source:
+  url: https://github.com/xtensor-stack/xtensor/archive/${{ version }}.tar.gz
+  sha256: f87259b51aabafdd1183947747edfff4cff75d55375334f2e81cee6dc68ef655
+
+build:
+  number: 0
+  skip:
+    # note that the value is a minijinja expression
+    - osx or win
+
+requirements:
+  build:
+    - ${{ compiler('cxx') }}
+    - cmake
+    - if: unix
+      value: make
+  host:
+    - xtl >=0.7,<0.8
+  run:
+    - xtl >=0.7,<0.8
+  run_constrained:
+    - xsimd >=8.0.3,<10
+
+test:
+  - script:
+      - if: unix
+        value:
+          - test -d ${PREFIX}/include/xtensor
+          - test -f ${PREFIX}/include/xtensor/xarray.hpp
+          - test -f ${PREFIX}/share/cmake/xtensor/xtensorConfig.cmake
+          - test -f ${PREFIX}/share/cmake/xtensor/xtensorConfigVersion.cmake
+      - if: win
+        value:
+          - if not exist %LIBRARY_PREFIX%\include\xtensor\xarray.hpp (exit 1)
+          - if not exist %LIBRARY_PREFIX%\share\cmake\xtensor\xtensorConfig.cmake (exit 1)
+          - if not exist %LIBRARY_PREFIX%\share\cmake\xtensor\xtensorConfigVersion.cmake (exit 1)
+  # compile a test package
+  - if: unix
+    then:
+      files:
+        - testfiles/cmake/*
+      extra_requirements:
+        build:
+          - ${{ compiler('cxx') }}
+          - cmake
+          - ninja
+      script: |
+        cd testiles/cmake/
+        mkdir build; cd build
+        cmake -GNinja ..
+        cmake --build .
+        ./target/hello_world
+
+  - downstream: xtensor-python
+  - downstream: xtensor-blas
+  - imports:
+      - xtensor_python
+      - xtensor_python.numpy_adapter
+```
+
+Or a multi-output package
+
+```yaml
+context:
+  name: mamba
+  libmamba_version: "1.4.2"
+  libmambapy_version: "1.4.2"
+  # can also reference previous variables here
+  mamba_version: ${{ libmamba_version }}
+  release: "2023.04.06"
+  libmamba_version_split: ${{ libmamba_version.split('.') }}
+
+# we can leave this out
+# package:
+#   name: mamba-split
+
+# this is inherited by every output
+source:
+  url: https://github.com/mamba-org/mamba/archive/refs/tags/${{ release }}.tar.gz
+  sha256: bc1ec3de0dd8398fcc6f524e6607d9d8f6dfeeedb2208ebe0f2070c8fd8fdd83
+
+build:
+  number: 0
+
+outputs:
+  - package:
+      name: libmamba
+      version: ${{ libmamba_version }}
+
+    build:
+      script:
+        script: ${{ "build_mamba.sh" if unix else "build_mamba.bat" }}
```

Comment:
```suggestion
      script: ${{ "build_mamba.sh" if unix else "build_mamba.bat" }}
```
Was this a mistake?

Also, on a related note, I've always found it confusing that `script:` can either be a string that is the name of a file which is a script or it can be a list of strings which will be converted into a bash or bat script. What can this new format do to make this less confusing?

Reply:
Yep, that was a mistake, thanks! 

I see two ways of achieving this, either by having another level in the dict, e.g.

```yaml
script:
  file: myfile.sh
  # or
  content: |
    ....
```

Or having an additional key, like `script_file` vs `script` or something like that.

Ok, maybe a third one would be something like: `script: ${{ embed("build.sh") }}` and write a jinja function that pulls in the contents of the file. This could be quite useful for other files as well (e.g. `README`, ...).

Reply:
I have to say that I like the current syntax a lot but I get that it is confusing.

Reply:
This could potentially allow choosing the interpreter to use:
```yaml
script:
  shell: bash
  file: build.sh
```

Reply:
> I have to say that I like the current syntax a lot but I get that it is confusing.

@beckermr Could you elaborate on why you like that `script:` is a keyword with two different meanings?

Reply:
I like not having to remember the different names for the different things.

Reply:
Actually my first thought when I saw `script: ${{ "build_mamba.sh" if unix else "build_mamba.bat" }}` was: "can't we have a default that abstracts away the different file extensions?"

Something like[^1]:
```
        script: build_mamba  # will use .sh on unix and .bat on win
```

If someone needs differently named scripts, they could always drop down to selectors, but it would make one more source of noise go away in the vast majority of cases.

[^1]: if we need to disambiguate from `script:` for compatibility (is this relevant in a new format that's already breaking things?), we could name the key `script_auto_ext:` or something like that.

Reply:
Actually the logic for that could be (since `build_mamba` / `build_mamba.sh` / etc. are just strings): If `$script + ".sh"` exists (on unix; `.bat` on windows) execute it, if it doesn't exist, assume the extension is already in the script, i.e. execute `$script`.